contracts/protocol/clients/voucher/BosonVoucher.sol

```diff
@@ -816,42 +805,28 @@ contract BosonVoucherBase is
         )
     {
         // Not committable if _committed already or if token has an owner
+
         if (!_committed[_tokenId] && !_exists(_tokenId)) {
-            // If are reserved ranges, search them
-            uint256 length = _rangeOfferIds.length;
-            if (length > 0) {
-                // Binary search the ranges array
-                uint256 low = 0; // Lower bound of search (array index)
-                uint256 high = length; // Upper bound of search
-
-                while (low < high) {
-                    // Calculate the current midpoint and get the offer id
-                    uint256 mid = (high + low) / 2;
-                    uint256 currentOfferId = _rangeOfferIds[mid];
-
-                    // Get the range stored at the midpoint
-                    Range storage range = _rangeByOfferId[currentOfferId];
-
-                    // Get the beginning of the range once for reference
-                    uint256 start = range.start;
-
-                    if (start > _tokenId) {
-                        // Split low and search again if target too high
-                        high = mid;
-                    } else if (start + range.length - 1 >= _tokenId) {
-                        // Is token in target's reserved range?
-
-                        if (start + range.minted - 1 >= _tokenId && _tokenId > range.lastBurnedTokenId) {
-                            // Has it been pre-minted and not burned yet?
-                            committable = true;
-                            offerId = currentOfferId;
-                            owner = range.owner;
-                        }
-                        break; // Found!
-                    } else {
-                        // No? It may be in a higher range
-                        low = mid + 1;
-                    }
+            // it might be a pre-minted token. Preminted tokens have offerId in the upper 128 bits
+            offerId = _tokenId >> 128;
+
+            if (offerId > 0) {
+                // Get the range stored at the midpoint
+                Range storage range = _rangeByOfferId[offerId];
+
+                // Get the beginning of the range once for reference
+                uint256 start = range.start;
+                // _tokenId = _tokenId & type(uint128).max;
+
+                if (
+                    start > 0 &&
+                    start <= _tokenId &&
+                    start + range.minted - 1 >= _tokenId &&
+                    _tokenId > range.lastBurnedTokenId
+                ) {
+                    // Has it been pre-minted and not burned yet?
```

Comment:
Yes `getBosonOffer` is rather costly method. It calls both `getExchange` and `getOffer` on the diamond. And because these are external calls everything needs to be retrieved from storage, copied to memory in send between contracts (around 30 storage slots).

But if you make a dedicated function `getOfferType(uint256 _offerId)` in the protocol that really returns only the type, I think it should be fine.